src/enumeration.jl

```diff
@@ -38,6 +38,19 @@ all_subtypes(ts::Vector, scfg :: SearchCfg, result :: Channel) = begin
             continue
         end
 
+        # Special case for Unions
+        # Since Unions are nested (`Union{A, B, C} === Union{A, Union{B, C}}`)
+        # we can peel the first case, substitute it for the Union, and add the
+        # rest of the Union to the worklist
+        i = findfirst(t -> t isa Union, tv)
```

Comment:
I guess this should do the job but I think it's a wrong place to put it. The smell of it is the `findfirst` call. You see, `all_subtypes` and `generate_subtypes` recur over each and every individual type anyway, so there shouldn't be a need for random access over those. Therefore, the right place to deal with unions, I think, is `generate_subtypes`. It's already recursive and it inspects every type, so when it sees a union it could recursively call itself twice: with the union substituted with the first component of that union and then with the second component of it.

Counterpoint to above is that I'm not sure how much allocation we do in either approach ("yours" and "mine").  It may be the case that doing it in `generate_subtypes` can spend more memory. I haven't thought it through, so can't say for sure.

Overall, I'm fine with merging it as is unless you feel like implementing the above idea. Please, let me know your thoughts.

Reply:
Sure, I can move it there!

Reply:
Awesome, thank you!